### ISSUE TITLE
Performance degradation of the X509Certificate2's private key constructor when running in Linux

### ISSUE THREAD

**Issue by thomasjohnson94** [dotnet/runtime#97753]
**Background**

We have a service that runs in .NET 8 using docker in Kubernetes, this service is responsible for sending/receiving files over HTTP. As part of this we support CMS Signing/Verification as well as Encryption/Decryption using public and private key certificates (X509Certificate2). For each transmission, we construct a certificate object to perform the signature verification/decryption using private keys when necessary.

**Description**

We've noticed a significant performance degradation in the X509Certificate2 constructor used for loading a private key certificate, which has caused a degradation of the service as a whole when processing a lot of requests in parallel.

We've tested the following code which constructs 1000 certificates serially on the following platforms;

- .NET 8 in docker using the ASP.NET image "mcr.microsoft.com/dotnet/aspnet:8.0"
- .NET 8 in docker using the nightly ASP.NET image
- .NET 6 in docker using the ASP.NET image "mcr.microsoft.com/dotnet/aspnet:6.0"

```
static void Main(string[] args)
{
	Console.WriteLine($"Running in [{RuntimeInformation.FrameworkDescription}]");

	byte[] arrBytes = File.ReadAllBytes(/*Path to private key cert*/);

	Stopwatch sw = Stopwatch.StartNew();

	for (int i = 0; i < 1000; i++)
	{
		_ = new X509Certificate2(arrBytes, /*pwd*/, X509KeyStorageFlags.Exportable);
	}

	Console.WriteLine(sw.ElapsedMilliseconds.ToString());

	arrBytes = File.ReadAllBytes(/*Path to public key cert*/);

	sw = Stopwatch.StartNew();

	for (int i = 0; i < 1000; i++)
	{
		_ = new X509Certificate2(arrBytes);
	}

	Console.WriteLine(sw.ElapsedMilliseconds.ToString());
}
```

We get around ~14000ms for private key certs when running in .NET 8, and ~100ms for public key certificates.
![image](https://github.com/dotnet/runtime/assets/150784560/3a200cdf-f212-45af-8298-f05fd06c34b5)

We get around ~2000ms for private key certs when running in .NET 6, and 10ms for public key certificates.
![image](https://github.com/dotnet/runtime/assets/150784560/c9c336f2-77b9-4311-8943-6c81a0b5fce0)

The OpenSsl version between the .NET 6 and .NET 8 image has also been updated - so it's entirely possible that OpenSsl is doing something different between the versions, which if that is the cause then we can close this off.

We've also worked around this by implementing a caching layer on top of our certificate access code, which has alleviated the issue.

**Comment by MichalPetryka** [comment]
Probably a duplicate of #97727?

**Comment by vcsjones** [comment]
> Probably a duplicate of #97727?

Very likely. I'll keep this one open until I can confirm.

**Comment by vcsjones** [comment]
This was indeed the same cause and fixed by https://github.com/dotnet/runtime/pull/97827.
